airflow/utils/python_virtualenv.py

```diff
@@ -35,12 +38,44 @@ def _generate_virtualenv_cmd(tmp_dir: str, python_bin: str, system_site_packages
     return cmd
 
 
-def _generate_pip_install_cmd(tmp_dir: str, requirements: List[str]) -> Optional[List[str]]:
+def _generate_pip_install_cmd(tmp_dir: str,
+                              requirements: List[str],
+                              connection_id: Optional[str] = None
+                              ) -> Optional[List[str]]:
     if not requirements:
         return None
-    # direct path alleviates need to activate
-    cmd = [f'{tmp_dir}/bin/pip', 'install']
-    return cmd + requirements
+
+    if connection_id:
+        con: Connection = BaseHook.get_connection(connection_id)
+        user = con.login
+        schema = con.schema or 'https'
+        password = con.get_password()
+        port = con.port or ''
+        host = con.host
+        path = con.extra_dejson.get('path', '')
+        pypi_as_fallback = con.extra_dejson.get('pypi_as_fallback', False)
+
+        if user:
+            netloc = f'{user}:{password}@{host}'
+        else:
+            netloc = host
+
+        netloc = netloc + f':{port}' if port else netloc
+
+        index_url = SplitResult(scheme=schema,
+                                netloc=netloc, path=path, query='', fragment='').geturl()
+
+        private_cmd = [f'{tmp_dir}/bin/pip',
+                       'install',
+                       f'--index-url', index_url]
+
+        if pypi_as_fallback:
+            private_cmd.extend([f'--extra-index-url', 'https://pypi.org/simple'])
```

Comment:
Se my comment here: https://github.com/apache/airflow/pull/16954#discussion_r669168968

If you must allow PyPI, set PyPI as the main index, not an extra index. This does not do what most people think it does.

Also I would avoid using the term “fallback” since it gives an impression PyPI is only accessed if a package is not available on the private index. That is wrong. Something like `allow_pypi` would be better (although I still strongly prefer not having this option at all).

Reply:
In my organisation we need for example to use both, I think sometime its parameter is required. I'll change this to the allow_pypi - it seems more straight forward.

Reply:
Any way else you suggest to do it ?

Reply:
In general, when your organisaion needs `--extra-index-url`, you don’t really need it and are identifying the issue wrong 🙂 There are exceptions, e.g. you are deploying to exotic hardware that PyPI doesn’t have wheels for, and you want to supply wheels for those PyPI project. For those exceptional use cases, setting `--extra-index-url` (while keeping the main index URL pointing to PyPI) is the right thing to do. But if what you need is to install a package not on PyPI, `--extra-index-url` is the wrong solution.

The suggestion is, instead of having multiple indexes serving different packages, making all requests go through the private index server you control. The index is configured in a way that when your private package is requeted, serve your own wheels, and when a PyPI package is requested, forward the request to the actual PyPI. This mechanism is implemented by Python index tools such as devpi (called index inheritance). I heard Artifactory can do that as well, although I’ve never personally understood how Artifactory works and can’t confirm.

Reply:
I m pretty sure its safe 'as implemented' now since its a flag that should be 'on' that means you know what you are doing.

Reply:
I don’t quite understand your last message, but what I can say is the PR, as it currently stands, does not provide a good approach to set `--extra-index-url`, and I would like Airflow to not encourage the approach.